Add functions to normalize and unnormalize by population (issue #69).

kjytay opened the pull request:
To address issue #69. Also incorporated into `production_forecaster`, and added other documentation. Some notes:

1. The code for the production forecaster is changed slightly to add this functionality, so while I don't expect it to break anything, we should probably test run before merging into `develop`.
2. I convert the population data into a named vector so that I can avoid a potentially costly `left_join()`.
3. There is no clean way to get `geo_type` from `production_forecaster`'s arguments, hence the somewhat unusual argument in the `invnorm_by_population()` call.
4. At the moment `normalize_by_population()` takes in a `df_list` while `invnorm_by_population()` takes in a `predict_df`. I've done it this way since these are the form the data will typically have when we call these functions. It is possible to standardize for both to take in a dataframe, but that means moving some of the checking and looping code from `normalize_by_population` into `production_forecaster`, which I would rather not do.

dajmcdon commented on forecasters/animalia/R/normalize_by_population.R:
```diff
@@ -0,0 +1,54 @@
+#' @return `df_list` but with the `value` column appropriately normalized.
+#' 
+#' @importFrom dplyr mutate select
+normalize_by_population <- function(df_list, norm_by_popn, base = 1e5) {
```
let's go ahead and export this function.

also, `norm_by_popn` is not a particularly meaningful name. How about `signals_to_normalize` or some such?

I'd also like to give it default values. I think you can do: `signals_to_normalize = rep(list(FALSE), nsigs)`. It will get evaluated lazily after `nsigs` is computed, and also avoids some error checking.

dajmcdon replied:
Sure. I missed that it would handle a length one logical. Can it deal with a vector of logicals rather than a list?

dajmcdon commented on forecasters/animalia/R/normalize_by_population.R:
```diff
@@ -0,0 +1,54 @@
+    }
+  }
+  return(df_list)
+}
```
empty line at the end

kjytay replied:
Done.

dajmcdon commented on forecasters/animalia/R/production_forecaster.R:
```diff
@@ -30,6 +30,12 @@
 #'   specified, defaults to the levels required by the COVID Forecast Hub.
 #' @param lambda vector of values to use for the regularization parameter
 #'   in quantile lasso
+#' @param norm_by_popn Should the response/features be normalized by population?
```
rename to be consistent with the above version

kjytay replied:
Renamed.

dajmcdon commented on forecasters/animalia/R/transformer.R:
```diff
@@ -1,3 +1,20 @@
+#'   `transform` is a list). Default is `NULL` for both `transform` and
+#'   `inv_trans`, which means no transformations are applied.
+#' 
+#' @return `df` but with the `value` columns appropriately transformed.
```
Thanks for documenting my function!

dajmcdon commented on forecasters/animalia/tests/testthat/test-population.R:
```diff
@@ -0,0 +1,109 @@
+state_geo_values <- c("ak", "al", "az")
+state_population <- c(731545, 4903185, 7278717)
+county_geo_values <- c("06037", "17031", "49035")
+county_population <- c(10039107, 5150233, 1160437)
```
unfortunately, I'm going to ask you to revise all of this test.

Testing with data out of the api is fairly unstable and results in lots of problems. Better is to use the mocking functions. Though here, I don't even think you need to do that. I would create some fake data in place of `state_df_list` and `county_df_list`. Here's [an example](https://github.com/cmu-delphi/covidcast/blob/a15fd603558a2288dd7cbebb9392436f560fd2c6/R-packages/evalcast/tests/testthat/test-evaluate_predictions.R#L4).

kjytay replied:
Revised.

kjytay commented on forecasters/animalia/R/invnorm_by_population.R:
```diff
@@ -0,0 +1,38 @@
+#' @return `df` but with the `value` column appropriately unnormalized.
+#' 
+#' @importFrom dplyr mutate select
+invnorm_by_population <- function(df, geo_type, base = 1e5) {
```
I can do the list processing, but it means that for the production forecaster, we have to coerce `prediction_df` into a list, call the function, then peel off the first element:
```
predict_df <- invnorm_by_population(
        list(predict_df), 
        geo_type = attr(df_list[[1]], "metadata")$geo_type)[[1]]
```

Would that be preferable over the current setup?

dajmcdon replied:
What I mean is, have `invnorm_by_population` take in an object. If it's not a list, convert it to one, with one element, then unnormalize elements or not as with `normalize_by_populaiotn()` (so we need another argument to determine), finally, return a list (if the input was a list) or a single data frame if it wasn't.

This way, suppose I have some covidcast signals, I can normalize or inverse normalize as I wish. We just use specific versions in production.

kjytay replied:
Fixed (subsumed under `normalize_by_population()`).

kjytay commented on forecasters/animalia/R/invnorm_by_population.R:
```diff
@@ -0,0 +1,38 @@
+    select(-population)
+  
+  return(df)
+}
```
Done.

kjytay commented:
As discussed, we want both normalization and unnormalization to work for a single data frame or a list of data frames. With this goal, there is so much overlap between the `normalize_by_population()` and `invnorm_by_population()` functions that I've deleted `invnorm_by_population()` and it is now implemented as `normalize_by_population(invert = TRUE)`.
